### ISSUE TITLE
[HL1/HL1SDK] Mouse input issues

### ISSUE THREAD

**Issue by RichardRohac** [ValveSoftware/halflife#1377]
Raw input works correctly on current (non beta) Half Life Steam build (just running regular Half Life), I get no lock on mouse rotation. But, compiled latest (8/10/2013) Half Life SDK (VS2010), I get the rotation lock when running the mod and raw input is enabled (same engine build is running, must be SDK issue I guess).

I also get no mouse wheel input whatsoever (every current GoldSrc game). (Microsoft Wireless Mouse 1000)

(Windows build)

**Comment by sixcentgeorge** [comment]
so any news ?
i played again my mod and saw that there is still the bug....while it is simple to correct : like i said a year ago , it is a question of path ...the mouse behave like in 1998 when the name is gearbox...but is "limited when the path is having an other name....
valve is having a bunch of lazies .....that make bad games and now kill others games..except black mesa the half-unfinished game that is a copy of hl made with "decomposition" of maps...
http://forums.steampowered.com/forums/showthread.php?t=3288444
yesterday i made a post in this topic with a link to http://www.consumerreports.org/cro/food/how-safe-is-your-ground-beef
that was silently removed...i hate forum that does these cleaning...they only tell lies and show the world as a paradise ...while it is a lot more like the "american burger" that has not only cows inside...
that is "funny" to see the difference between hl2 and hl3...the first was asked every day while the second interests only nobody..

**Comment by thopvna** [comment]
You can try: mouse & touchpad -> disable scroll inactive windows when i hover over them

**Comment by djdallmann** [comment]
Doesn't SDL_GetRelativeMouseState get the last state since the previous mouse update and not current mouse position which then causes mouse movement to be off by one update or am I misunderstanding this SDL function?

**Comment by SamVanheer** [comment]
See #1900

**Comment by RichardRohac** [comment]
@mikela-valve Seeing that some work is being done on the engine again, could this be looked into?

I experience the same mouse issues with the current engine version. 

@SamVanheer This is all great, but the actual SDL build needs to be updated as well.
The SDL version that comes with the engine is too old to handle wheel scrolling quirks of some (Microsoft) mouse types. This was solved some time ago (http://hg.libsdl.org/SDL/rev/d8fb783475d5).

**Comment by SamVanheer** [comment]
Yeah the SDL2 library needs updates for a couple reasons, see also #1887

**Comment by djdallmann** [comment]
Yes, plz update SDL

**Comment by jirikivaari** [comment]
Yeah some mouuse wheel issues were fixed in another mod by replacing the SDL2 in HL1 folder.

**Comment by sinisterseed** [comment]
It's been said and stressed before as far as I can see, but this does indeed need to be fixed as soon as possible as input in one of the most crucial elements of a game. I've been revisiting some mods recently, namely They Hunger, and using Raw Input is simply impossible in its current state, just locks me up/down and the view starts spinning uncontrollably. I'm on W10 version 1909.

I'm quite sensitive to input lag myself, so being forced to not use it has made my experience almost borderline unplayable. It feels as if the mouse is almost skipping frames at times, very jittery and imprecise. For some reason raising the framerate cap and using fps_override has made it more bearable, but naturally the core issue is still there, just less noticeable. I'm quite glad to see the GoldSrc games still receive love after all these years, but at the same time I am perplexed at the fact that such a major issue is still yet to be properly addressed.

**Comment by HalfWayLambda** [comment]
@mikela-valve Could you take a look at this issue?

**Comment by djdallmann** [comment]
Would really like to see this get fixed :(

**Comment by sixcentgeorge** [comment]
@mikela-valve can you make the two "mod" 's dlls for a mod i am making .. or give me somebody that can do it .. i am ready to give a little amount of money .
i want something basic like the gearbox engine dlls . i asked at gearbox steam community but no one is there except "customers having bugs" . how to get the source code ? or how to buy the "commercial hlsdk "  ?

for those having the bug.. they should test using a well known steam path : like gearbox or dmc...
it is an added "end of life" error to force people uses the source "engine"  that no one likes so much..

**Comment by Solokiller** [comment]
@mikela-valve i looked into this and i think the solution is as Alfred said above to add a way to hint the engine that SDL is being used.

Currently the engine sets relative mouse mode using `SDL_SetRelativeMouseMode` if `BUsesSDLInput` returns true. `BUsesSDLInput` only considers official Valve games using an app id and client dll path check. If the client could hint to the engine that it's using SDL2, as any Half-Life mod using the current SDK is doing, then the engine can correctly set up SDL state as it does for official Valve games.

To test this i hex edited the engine to replace the filename for `gearbox\cl_dlls\client` to `updated\cl_dlls\client`. When i ran my mod it no longer had the issue. Mouse sensitivity did go down, probably as a result of the mouse data being received by the client being different.

It's probably easier to test by putting a mod dll in place of an official game dll, like `valve\cl_dlls\client.dll`. But this confirms that it works with files that don't have the path of an official game dll.

A new engine function could be used to provide the hint:
```
//In APIProxy.h
	pfnEngSrc_pfnVguiWrap2_GetMouseDelta_t pfnVguiWrap2_GetMouseDelta;
    void (*SetClientUsesSDL)(int usesSDL);
} cl_enginefunc_t;
```

Then on startup, probably in `Initialize` the client tells the engine:
```
//In cdll_int.cpp, function Initialize
memcpy(&gEngfuncs, pEnginefuncs, sizeof(cl_enginefunc_t));

gEngfuncs.SetClientUsesSDL(1);
```

The value provided would be used by `BUsesSDLInput`. If the client does not tell the engine it should default to identifying the client as it does now.

So the engine should perform these steps:
* Initialize global `int g_bUseRawInput` to `-1`
* Load and initialize client dll
* Check if `g_bUseRawInput` is either `0` or `1`
* If not, check if it's an official game using existing logic
* `g_bUseRawInput` should now be either 0 or 1

In code:
```
int g_bUseRawInput = -1;

qboolean BUsesSDLInput()
{
    return g_bUseRawInput != 0;
}

void CL_SetClientUsesSDL(int useSDL)
{
    g_bUseRawInput = useSDL != 0;
}

void DetermineSDLInput()
{
    if (g_bUseRawInput != 0 && g_bUseRawInput != 1)
    {
        g_bUseRawInput = 0;

        if (BIsValveGame())
        {
            g_bUseRawInput = 1;
        }
        else if (Q_strstr(g_szfullClientName, "valve/cl_dlls/client"))
        {
            g_bUseRawInput = 1;
        }
        //Remaining filename checks here
    }
}
```

**Comment by SamVanheer** [comment]
Addendum: the way the engine detects whether a game uses SDL causes it to hide the mouse cursor in VGUI1 menus when `m_rawinput` is 0. I encountered this issue in Half-Life: Opposing Force Updated.

**Comment by djdallmann** [comment]
Is it possible to make HL use newer SDL2 without breaking hl or potentially getting vac banned? File link, replace file (steam.exe -noverifyfiles).

**Comment by SamVanheer** [comment]
If you run the game with `-insecure` then VAC won't be enabled when running listen servers but playing on VAC secure servers will still carry the risk of getting banned.
